sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```
@@ -0,0 +1,1121 @@
+       */
+      private <T> T cloneIfPossiblyMutable(T input, Coder<T> coder) throws CoderException,
+          ClassCastException {
+        // If the input object is not of known immutable type, clone the object.
+        if (!isKnownImmutable(input)) {
+          input = CoderUtils.clone(coder, input);
+        }
+        return input;
+      }
+
+      /**
+       * Utility method to check if the passed object is of a known immutable type.
+       */
+      private boolean isKnownImmutable(Object o) {
+        Set<Class<?>> immutableTypes = new HashSet<Class<?>>(
+            Arrays.asList(
+                String.class,
+                Byte.class,
+                Short.class,
+                Integer.class,
+                Long.class,
+                Float.class,
+                Double.class,
+                Boolean.class,
+                BigInteger.class,
+                BigDecimal.class));
+        return immutableTypes.contains(o.getClass());
+      }
+
+      @Override
+      public void close() throws IOException {
+        LOG.info("Closing reader after reading {} records.", recordsReturned);
+        if (currentReader != null) {
+          currentReader.close();
+          currentReader = null;
+        }
+      }
+
+      @Override
+      public Double getFractionConsumed() {
+        if (doneReading) {
+          return 1.0;
+        }
+        if (currentReader == null || recordsReturned == 0) {
+          return 0.0;
+        }
+        return progressValue.doubleValue();
+      }
+
+      /**
+       * Returns RecordReader's progress.
+       */
+      private Double getProgress() {
+        try {
+          synchronized (currentReader) {
+            return (double) currentReader.getProgress();
+          }
+        } catch (IOException | InterruptedException e) {
+          LOG.error(HadoopInputFormatIOConstants.GETFRACTIONSCONSUMED_ERROR_MSG + e.getMessage(), e);
+          return 0.0;
```

hrm... I don't think we should suppress the exception here. Is there a reason to?

---

Removed the suppressing code and instead we are now throwing an IOException with a message and throwable object.